rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
```

Comment:
We have had problems with path params in the past, specifically that path
params are not encoded automatically by libraries as they are for query
params.  So I see some risk there.  Also, with a path param you have to
pick the order for those parameters.  What is intuitive to you, is not
intuitive to me.  Using a query param means order doesn't matter and I have
a more consistent experience.  How would your example account for searching
for a user by login?  What if my login was a number and that number was
also someone's ID?  Ambiguity is a problem that goes away with query
params, you have to explicitly identify with a parameter name all the
values you want in your request.

On Fri, Jun 24, 2016 at 8:58 AM, dewayneHat <notifications@github.com>
wrote:

> In rest_specification/restful_service_specification.md
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r68393921>
> :
>
> > +| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
> > +| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
> > +| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
> > +| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
> > +| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
> > +
> > +Endpoints may not support all methods (i.e. a service may provide read-only
> > +access to a resource, or a service may not support certain operations for
> > +performance or other considerations; for example, `GET` may not be supported
> > +for `user` if listing all users is disallowed for performance reasons.
> > +
> > +## Operations Against Specific Resources
> > +Specific resources will be addressed through the base URI plus the identifier
> > +for that resource.
> > +
> > +    /user/{id}
>
> Another option would be to require that every resource has a single id,
> but this might be more controversial. I like the simplicity of that, but
> sometimes it forces you to leave the natural domain/semantics of the
> resource (imagine a case where business users always talk about identity of
> a domain object in terms of two values, but now we require it's determined
> by an id, which means we have 2 ways of getting at the same thing, etc...
> seems maybe messy).
>
> IMO, APIs are not for business users, they are for developers and the
> services/applications they write. If using a single id to identify
> resources that have compound ids makes the API more useable for the
> developer writing code against it, then that is what we should do. The
> business can continue to think of it their way without knowing the
> implementation details of the API.
>
> /users/{id}
> I think single id is probably the api that will end up being the most
> intuitive and natural to use as a consumer of the API. However, this could
> lead to situations where a consumer of the API may have the compound id of
> the resource, and not the "api id". So we might end up creating an api to
> support that as well anyhow.
>
> For the sake of example let's say we recently merged users from several
> different databases into a single db. And for some reason, we decided to
> make the user's compound id: 'id' and 'dborigin'. Where dborigin is the
> database where the user originated from (0=Oracle, 1=mysql, 2=mongo). Then
> the api could look something like this
> GET /users -- Returns all users (Paginated of course)
> GET /users/1 -- Returns all users with id 1 (Max of 3)
> GET /users/1/dborigin/0 -- Returns single user with id 1 from Oracle db
>
> I know this isn't perfect, but it makes sense intuitively.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r68393921>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAdUhJGaWWp4MvkrO-W6pX-3rnheUXDDks5qO9RngaJpZM4IopeE>
> .
>